java/vector/src/main/codegen/templates/UnionVector.java

```diff
@@ -586,7 +686,9 @@ public ValueVector getVectorByType(int typeId) {
   }
 
     public ValueVector getVectorByType(int typeId, ArrowType arrowType) {
-      switch (MinorType.values()[typeId]) {
+      Types.MinorType type = (typeIds[typeId] != null) ?
+          Types.getMinorTypeForArrowType(typeIds[typeId].getType()) : Types.MinorType.values()[typeId];
+      switch (type) {
```

Comment:
Here we are still relying on the MinorType ordinal as the type ID?
If so, the sparse union vector is still not aligning with the specification, and that causes the integration test to fail?

Reply:
This switch uses the `typeIds` array to extract the MinorType associated with the logical type. If it does not exist in the array (usually because a logical type wasn't previously assigned) it defaults to the id of the minor type. So Java always uses the MinorType as the type id and allows c++ (and others) to send arbitrary type mappings.